File: ealvalog-android/src/androidTest/java/ealvalog/impl/AndroidLoggerTest.java
/*
 * Copyright 2017 Eric A. Snell
 *
 * This file is part of eAlvaLog.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ealvalog.impl;

import ealvalog.LogLevel;
import ealvalog.Logger;
import ealvalog.Loggers;
import ealvalog.Marker;
import ealvalog.util.LogMessageFormatter;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.when;

import android.support.test.runner.AndroidJUnit4;
import android.util.Log;

@SuppressWarnings("deprecation") @RunWith(AndroidJUnit4.class)
public class AndroidLoggerTest {
  private static final String TAG = "AndroidLoggerTest";
  @SuppressWarnings("WeakerAccess") @Mock LogHandler mockHandler;

  @Before
  public void setup() {
    MockitoAnnotations.initMocks(this);
    AndroidLogger.setHandler(mockHandler);
    Loggers.setFactory(new AndroidLoggerFactory());
  }

  @Test
  public void testPrintLog() {
    when(mockHandler.isLoggable(TAG, Log.ASSERT)).thenReturn(true);

    //given
    final Logger logger = Loggers.get();
    final String msg = "The msg";
    logger.log(LogLevel.CRITICAL, msg);

    then(mockHandler).should(atLeastOnce()).prepareLog(eq("AndroidLoggerTest"),
                                                       same(Log.ASSERT),
                                                       isNull(Marker.class),
                                                       isNull(Throwable.class),
                                                       isNull(StackTraceElement.class),
                                                       any(LogMessageFormatter.class),
                                                       eq(msg));
  }

  @Test
  public void testPrintLogHandlerWantsLocation() {
    when(mockHandler.shouldIncludeLocation(TAG, Log.ASSERT, null, null)).thenReturn(true);
    when(mockHandler.isLoggable(TAG, Log.ASSERT)).thenReturn(true);

    //given
    final Logger logger = Loggers.get();
    final String msg = "The msg";
    logger.log(LogLevel.CRITICAL, msg);

    then(mockHandler).should(atLeastOnce()).prepareLog(eq("AndroidLoggerTest"),
                                                       same(Log.ASSERT),
                                                       isNull(Marker.class),
                                                       isNull(Throwable.class),
                                                       any(StackTraceElement.class),
                                                       any(LogMessageFormatter.class),
                                                       eq(msg));
  }

  @Test
  public void testPrintLogSetIncludeLocation() {
    when(mockHandler.isLoggable(TAG, Log.ASSERT)).thenReturn(true);

    //given
    final Logger logger = Loggers.get();
    logger.setIncludeLocation(true);
    final String msg = "The msg";
    logger.log(LogLevel.CRITICAL, msg);

    then(mockHandler).should(atLeastOnce()).prepareLog(eq("AndroidLoggerTest"),
                                                       same(Log.ASSERT),
                                                       isNull(Marker.class),
                                                       isNull(Throwable.class),
                                                       any(StackTraceElement.class),
                                                       any(LogMessageFormatter.class),
                                                       eq(msg));
  }

}

File: ealvalog-android/src/androidTest/java/com/ealva/ealvalog/android/AndroidLoggerTest.kt
/*
 * Copyright 2017 Eric A. Snell
 *
 * This file is part of eAlvaLog.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ealva.ealvalog.android

import android.util.Log
import androidx.test.runner.AndroidJUnit4
import com.ealva.ealvalog.LogLevel
import com.ealva.ealvalog.Loggers
import com.ealva.ealvalog.Markers
import com.ealva.ealvalog.core.ExtLogRecord
import com.ealva.ealvalog.java.JLoggers
import com.nhaarman.expect.expect
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.lang.RuntimeException

/**
 * Created by Eric A. Snell on 9/29/18.
 */
@RunWith(AndroidJUnit4::class)
class AndroidLoggerTest {
  @Before
  fun setup() {
    AndroidLoggerFactory.clearLoggerCache()
    Loggers.setFactory(AndroidLoggerFactory)
  }

  @Test
  fun testPrintLog() {
    // Given
    val mockHandler = mock<LogHandler> {
      on { isLoggable(TAG, LogLevel.CRITICAL, null, null) }.thenReturn(true)
      on { shouldIncludeLocation(TAG, Log.ASSERT, null, null) }.thenReturn(false)
    }
    AndroidLogger.setHandler(mockHandler)
    val logger = JLoggers.get(AndroidLoggerTest::class.java)
    val msg = "The msg"

    // When
    logger.log(LogLevel.CRITICAL, msg)

    // Then
    val recordCaptor = argumentCaptor<ExtLogRecord>()
    verify(mockHandler, times(1)).prepareLog(recordCaptor.capture())
    val record = recordCaptor.firstValue
    expect(record.loggerName).toBe(AndroidLoggerTest::class.java.name)
    expect(record.logLevel).toBe(LogLevel.CRITICAL)
    expect(record.message).toBe(msg)
    expect(record.thrown).toBeNull()
    expect(record.marker).toBeNull()
    expect(record.location).toBeNull()
  }

  @Test
  fun testPrintLogHandlerWantsLocation() {
    // Given
    val ex = RuntimeException("MyException")
    val mockHandler = mock<LogHandler> {
      on { isLoggable(TAG, LogLevel.WARN, null, ex) }.thenReturn(true)
      on { shouldIncludeLocation(TAG, Log.WARN, null, ex) }.thenReturn(true)
    }
    AndroidLogger.setHandler(mockHandler)
    val logger = JLoggers.get(AndroidLoggerTest::class.java)
    val msg = "The msg"

    // When
    logger.log(LogLevel.WARN, ex, msg)

    // Then
    val recordCaptor = argumentCaptor<ExtLogRecord>()
    verify(mockHandler, times(1)).prepareLog(recordCaptor.capture())
    val record = recordCaptor.firstValue
    expect(record.loggerName).toBe(AndroidLoggerTest::class.java.name)
    expect(record.logLevel).toBe(LogLevel.WARN)
    expect(record.message).toBe(msg)
    expect(record.thrown).toBe(ex)
    expect(record.marker).toBeNull()
    expect(record.location).toNotBeNull()
  }

  @Test
  fun testPrintLogSetIncludeLocation() {
    // Given
    val marker = Markers["MarkerName"]
    val mockHandler = mock<LogHandler> {
      on { isLoggable(TAG, LogLevel.ERROR, marker, null) }.thenReturn(true)
      on { shouldIncludeLocation(TAG, Log.ERROR, marker, null) }.thenReturn(false)
    }
    AndroidLogger.setHandler(mockHandler)
    val logger = JLoggers.get(AndroidLoggerTest::class.java)
    logger.includeLocation = true
    val msg = "The msg"

    // When
    logger.log(LogLevel.ERROR, marker, msg)

    // Then
    val recordCaptor = argumentCaptor<ExtLogRecord>()
    verify(mockHandler, times(1)).prepareLog(recordCaptor.capture())
    val record = recordCaptor.firstValue
    expect(record.loggerName).toBe(AndroidLoggerTest::class.java.name)
    expect(record.logLevel).toBe(LogLevel.ERROR)
    expect(record.message).toBe(msg)
    expect(record.thrown).toBeNull()
    expect(record.marker).toBe(marker)
    expect(record.location).toNotBeNull()
  }

  companion object {
    const val TAG = "AndroidLoggerTest"
  }
}
